packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,213 @@
+
+const WatermarkValueEditor = observer(
+  (props: {
+    watermarkState: QueryBuilderWatermarkState;
+    queryBuilderState: QueryBuilderState;
+  }) => {
+    const { watermarkState } = props;
+    const variable = watermarkState.value as VariableExpression;
+    const graph = watermarkState.queryBuilderState.graphManagerState.graph;
+    const parameterType = watermarkState.value?.genericType?.value.rawType;
+
+    const handleDrop = useCallback(
+      (item: QueryBuilderParameterDragSource): void => {
+        watermarkState.setValue(item.variable.parameter);
+      },
+      [watermarkState],
+    );
+
+    const [{ isParameterValueDragOver }, dropTargetConnector] = useDrop<
+      QueryBuilderParameterDragSource,
+      void,
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
+            // Doing a type check, which only allows dragging and dropping parameters of the same type or of child types
+            itemType &&
+            parameterType &&
+            (isSuperType(parameterType, itemType) ||
+              parameterType.name === itemType.name)
+          ) {
+            handleDrop(item);
+          } // prevent drop event propagation to accomondate for nested DnD
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <>
+        {watermarkState.value && (
+          <div key={variable.name} className="panel__content__form__section">
+            <div className="panel__content__form__section__header__label">
+              {variable.name}
+            </div>
+            <div className="query-builder__parameter-editor">
+              <PanelDropZone
+                isDragOver={isParameterValueDragOver}
+                dropTargetConnector={dropTargetConnector}
+              >
+                <BasicValueSpecificationEditor
+                  valueSpecification={guaranteeNonNullable(
+                    watermarkState.value,
+                  )}
+                  setValueSpecification={(val: ValueSpecification): void => {
+                    watermarkState.setValue(val);
+                  }}
+                  graph={graph}
+                  obseverContext={
+                    watermarkState.queryBuilderState.observableContext
+                  }
+                  typeCheckOption={{
+                    expectedType: graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                    match:
+                      parameterType ===
+                      graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                  }}
+                  resetValue={(): void => watermarkState.setValue(undefined)}
```

Comment:
I think resetValue should probably reset the value spec to the default value which I think we can make `watermarkValue`

Reply:
Everytime the value spec editor is changed though the value becomes the watermark value, so if I changed it to `watermarkState.setValue(watermarkState.value)` I believe a difference never occurs? 🤔